### ISSUE TITLE
Remote-process injection via Java agent (attach/detach)

### ISSUE THREAD

**Issue by rock3r** [rock3r/spectre#153]
## Background

Spectre currently supports two ways to drive a Compose Desktop UI:

1. **In-process** (`ComposeAutomator.inProcess()`) — test and UI share a JVM.
2. **HTTP transport** (`ComposeAutomator.http(...)` / `installSpectreRoutes(...)`) — test and UI in separate JVMs, but the hosting JVM must opt in at startup by mounting Ktor routes.

Both require the host JVM to wire up automation at startup. There is no way to **late-bind** to an already-running Spectre-instrumented JVM. This blocks several real workflows:

- Driving a long-running Compose Desktop app that was started without the routes wired up — typical exploratory/interactive use.
- Driving an IntelliJ-hosted Compose surface (Jewel tool window) from a separate inspector/test process, without modifying the plugin's startup.
- Future CLI tooling (`spectre attach <pid>`) that drives any Spectre-aware process.

The JVM **Attach API** (`com.sun.tools.attach.VirtualMachine`) lets one JVM load a Java agent into another running JVM. This is the standard mechanism profilers, coverage collectors, and IDE debuggers use. Spectre should build on it.

---

## Scope (v1)

**Targets in scope** — Spectre-instrumented JVMs only:
- Standalone Compose Desktop apps that include `dev.sebastiano.spectre:core` (and the new agent runtime artifact) on their classpath.
- IntelliJ-hosted Compose surfaces inside a Spectre-aware IntelliJ plugin (the `sample-intellij-plugin` shape — Spectre is loaded by the plugin's `PluginClassLoader`).

**Out of scope for v1:**
- Third-party Compose Desktop apps without Spectre on classpath. The agent does not inject Spectre's core — it bootstraps an existing copy.
- Remote-machine attach. Local IPC only.
- Windows. macOS + Linux only via Unix Domain Sockets; Windows tracked as a follow-up.
- Long-poll / streaming wire ops (`waitForVisualIdle`, idling resources, `withTracing`). Tracked as a follow-up.
- JDI / JVMTI alternative transports. Considered and rejected for v1; see the plan's "Rejected alternatives" section.
- Byte-code instrumentation / class retransformation. The agent bootstraps only.
- Auth / TLS on the IPC. UDS filesystem permissions are the trust boundary.
- Publishing the new module to Maven Central. v1 stays unpublished.

---

## Design summary

Full design is in [.plans/2026-05-22-issue-153-agent-attach-workshop.md](https://github.com/rock3r/spectre/blob/feature/issue-153-agent-attach/.plans/2026-05-22-issue-153-agent-attach-workshop.md) (workshop-mode plan, 11 tracer-slice milestones). Headlines:

### Thin agent + reflective classloader bootstrap

The agent JAR ships **only** bootstrap + IPC server code + transport DTOs + kotlinx-serialization. **No** Compose, Skiko, Kotlin stdlib, or `dev.sebastiano.spectre.core.*` classes. These resolve from the target's classpath at runtime.

On attach, the agent:
1. Scans `Instrumentation.getAllLoadedClasses()` for `dev.sebastiano.spectre.core.ComposeAutomator`.
2. Captures that class's `ClassLoader` (system loader for standalone; `PluginClassLoader` for IntelliJ plugins).
3. Reflectively loads `dev.sebastiano.spectre.agent.runtime.AgentBootstrap` through that loader.
4. `AgentBootstrap.start(udsPath)` runs the IPC server in that loader's namespace — so `window is ComposeWindow` checks resolve against the target's actual classes, not the agent's.

A disambiguation rule handles the edge case of multiple Spectre copies on the loader graph (prefer the `PluginClassLoader`-rooted candidate; fail loudly with `AmbiguousSpectreClasspathException` if no winner emerges).

### Wire transport: UDS + length-prefixed CBOR

Not HTTP. Not Ktor. The local-attach use case doesn't need a network listener:

- **Unix Domain Sockets** via `java.net.UnixDomainSocketAddress` (JDK 16+, no third-party dep).
- **Length-prefixed CBOR** frames over the stream, via `kotlinx-serialization-cbor`.
- Path includes attacker PID + UUID; agent registers a shutdown hook to unlink on JVM exit.
- Path passed to the agent via `agentArgs` (no `getSystemProperties` round-trip, no temp-file polling).

Trust boundary is filesystem permissions on the UDS path (mode 0600, owner-only). Sufficient for local-only.

### Three artifacts from `:agent`

1. **`spectre-agent`** — attach-side library for the client/test JVM. Contains `AgentAttach`, `AttachedAutomator`, `AttachOptions`, `SpectreProcesses`, the IPC client.
2. **`spectre-agent-runtime-all`** — the fat agent JAR with `Agent-Class` / `Premain-Class` manifest entries. The thing that gets `loadAgent`-ed. Tiny; bootstrap + IPC server + DTOs + serialization only.
3. **`spectre-agent-runtime`** — non-fat runtime library the target app puts on its compile/runtime classpath so the agent's reflective lookups find Spectre classes.

### Detach with dual cleanup paths

**Path A — explicit `AttachedAutomator.close()`** (normal):
1. Sends `Detach` op over the wire.
2. Agent drains in-flight requests, stops accepting new ones, closes the IPC server, releases the in-target `ComposeAutomator`, unlinks the UDS path, removes its shutdown hook.
3. Attacher calls `VirtualMachine.detach()`.

**Path B — JVM shutdown hook** (crash safety): target-side hook performs the same cleanup if Path A didn't run. No-op if already cleaned up.

Post-detach invariants verified in tests:
- No Spectre threads remain in the target.
- UDS path no longer exists.
- System property markers cleared.

### Client API sketch

```kotlin
@ExperimentalSpectreAgentApi
public object AgentAttach {
    public fun attach(pid: Long, options: AttachOptions = AttachOptions()): AttachedAutomator
}

@ExperimentalSpectreAgentApi
public class AttachedAutomator(
    public val automator: ComposeAutomator,
    public val pid: Long,
) : AutoCloseable

@ExperimentalSpectreAgentApi
public data class AttachOptions(
    public val agentJarPath: Path = resolveDefaultAgentJar(),
    public val udsPath: Path? = null,                 // null = auto-generated
    public val attachTimeoutMs: Long = 5_000,
    public val strictAgentLoadingCheck: Boolean = true,
)

@ExperimentalSpectreAgentApi
public object SpectreProcesses {
    public fun listJvmProcesses(): List<JvmProcessInfo>
    public fun findByName(nameFilter: String): List<JvmProcessInfo>
}
```

### JEP 451 UX

Per [JEP 451](https://openjdk.org/jeps/451), JDK 21+ prints a warning when an agent is dynamically loaded unless the target was started with `-XX:+EnableDynamicAgentLoading`. Future JDKs will disallow dynamic loading entirely.

Spectre's UX:
- `AttachOptions(strictAgentLoadingCheck = true)` (default) — detects via `VirtualMachine.getSystemProperties()` whether the flag is set; throws `AgentLoadingNotEnabledException` if not, with actionable guidance.
- `strictAgentLoadingCheck = false` — attach proceeds with a logged WARN matching the target's own stderr warning.
- For long-term-stable use, docs prescribe loading the agent at startup via `-javaagent:spectre-agent-runtime-all.jar=...` rather than dynamic attach.

### Attach API preconditions

The plan covers JDK Attach API failure modes that the original issue scope missed:
- Attacher must be a JDK (not a JRE) with the `jdk.attach` module.
- Same-UID requirement on POSIX (`ProcessHandle.info().user()` checked pre-flight).
- hsperfdata visibility limits for `VirtualMachine.list()`.
- `/tmp/.java_pid<pid>` reachability on hardened Linux containers.

Each surfaces as a specific exception subtype with diagnostic context.

---

## Milestone tracer slices

Workshop-mode plan defines 11 milestones, each producing an end-to-end demonstrable capability:

- **M-1** — Bootstrap spike: thin agent locates Spectre on target classpath, prints window count from a running `:sample-desktop:run`. De-risks the classloader story before any other work.
- **M-2** — Wire protocol scaffold: CBOR + length-prefixed framing, in-process round-trip.
- **M-3** — UDS transport + first real op: `windows()` over the wire against sample-desktop.
- **M-4** — Read ops: `allNodes()`, `findByTestTag()`, `screenshot()`.
- **M-5** — Write ops + detach: `click()`, `typeText()`, full Path A detach with post-detach invariants.
- **M-6** — IntelliJ-hosted Compose validation (manual spike; de-scope path documented if `PluginClassLoader` discovery proves intractable).
- **M-7** — JUnit `ComposeAutomatorAttachExtension` + new `:agent-test-fixture` module + shared `ChildJvmProcess` helper.
- **M-8** — End-to-end against `:sample-desktop` spawned as a child JVM (with a system-property-gated readiness sentinel).
- **M-9** — End-to-end against `:sample-intellij-plugin` via `intellij-ide-starter` (with de-scope path if M-6 hit its).
- **M-10** — Docs (`docs/guide/agent.md`), `mkdocs.yml`, `SECURITY.md` IPC trust-model entry, `STABILITY.md` opt-in entry, README update.
- **M-11** — Final `./gradlew check` green + repeat attach/detach cycle test.

Each milestone has its own TDD red/green cycle, baseline/end markers, and is gated by a plan-adherence review before advancing.

---

## Acceptance criteria

- New `:agent` Gradle module follows project conventions (kotlinJvm + detekt + ktfmt + JDK 21 toolchain + explicitApi).
- `./gradlew :agent:shadowJar` produces the fat agent JAR with the correct manifest, no shaded Compose / Skiko / Kotlin stdlib classes inside (verified by a JAR-contents assertion task).
- `AttachedAutomator.close()` performs full cleanup; ≥3 attach/detach cycles against the minimal fixture leave no Spectre threads or orphan UDS files.
- Same cycle passes against `:sample-desktop` (M-8). IntelliJ path (M-9) passes against `:sample-intellij-plugin` or is documented as a known v1 gap if R-1 materialized.
- `strictAgentLoadingCheck = true` throws on a target without `-XX:+EnableDynamicAgentLoading`; `false` proceeds with a WARN.
- All new public API under `dev.sebastiano.spectre.agent.*` requires `@ExperimentalSpectreAgentApi` opt-in.
- `docs/guide/agent.md` covers usage, `AttachOptions`, platform prerequisites, trust model; `mkdocs build --strict` passes.
- `./gradlew check` green.

---

## References

- Plan: [.plans/2026-05-22-issue-153-agent-attach-workshop.md](https://github.com/rock3r/spectre/blob/feature/issue-153-agent-attach/.plans/2026-05-22-issue-153-agent-attach-workshop.md)
- Worktree: `feature/issue-153-agent-attach`
- [JEP 451: Prepare to Disallow the Dynamic Loading of Agents](https://openjdk.org/jeps/451)
- JDK Attach API: `com.sun.tools.attach.VirtualMachine` (`jdk.attach` module)
- Existing HTTP cross-JVM transport: `server/` module, `docs/guide/cross-jvm.md` (this issue does NOT extend it)
- Architecture invariants: `docs/ARCHITECTURE.md`
- Security risk register: `docs/SECURITY.md`

**Comment by rock3r** [comment]
**Design pivot logged.** The issue body has been rewritten to match the Phase 1 workshop plan ([.plans/2026-05-22-issue-153-agent-attach-workshop.md](https://github.com/rock3r/spectre/blob/feature/issue-153-agent-attach/.plans/2026-05-22-issue-153-agent-attach-workshop.md), on branch `feature/issue-153-agent-attach`).

Three major shifts from the original issue:

1. **No HTTP transport.** Use Unix Domain Sockets + length-prefixed CBOR. No Ktor in the agent runtime.
2. **Thin agent.** Target must already have Spectre on its classpath. The agent ships only bootstrap + IPC + DTOs + serialization — no Compose/Skiko/stdlib shading. Solves the cross-classloader `is ComposeWindow` problem that would have silently broken window discovery.
3. **Scope narrowed.** v1 covers Spectre-instrumented standalone apps + IntelliJ-hosted Compose only. Third-party apps, Windows, JDI/JVMTI alternatives, and long-poll wire ops are explicit follow-ups.

The original issue body had the right shape but the wrong technical specifics. The plan's "Rejected alternatives" section explains why HTTP/Shadow-relocation/JDI/JVMTI were considered and dropped.
